modules/platform/nuxeo-platform-web-common/src/main/java/org/nuxeo/ecm/platform/web/common/idempotency/NuxeoIdempotentFilter.java

```diff
@@ -0,0 +1,196 @@
+
+    protected static final int MAX_CONTENT_SIZE = 1024 * 1024 * 5; // 5 MB
+
+    protected static final Set<String> IDEMPOTENT_METHODS = Set.of(
+            // safe methods according to RFC 7231 4.2.1
+            HttpGet.METHOD_NAME, HttpHead.METHOD_NAME, HttpOptions.METHOD_NAME, HttpTrace.METHOD_NAME,
+            // idempotent methods according to RFC 7231 4.2.2
+            HttpPut.METHOD_NAME, HttpDelete.METHOD_NAME);
+
+    protected Duration getTTL() {
+        ConfigurationService cs = Framework.getService(ConfigurationService.class);
+        if (cs != null) {
+            return cs.getDuration(TTL_DURATION_PROPERTY, DEFAULT_TTL);
+        }
+        return DEFAULT_TTL;
+    }
+
+    protected String getStoreName() {
+        ConfigurationService cs = Framework.getService(ConfigurationService.class);
+        if (cs != null) {
+            return cs.getString(STORE_PROPERTY, DEFAULT_STORE);
+        }
+        return DEFAULT_STORE;
+    }
+
+    @Override
+    protected void doFilter(HttpServletRequest request, HttpServletResponse response, FilterChain chain)
+            throws IOException, ServletException {
+        if (!doFilterIdempotent(request, response, chain)) {
+            chain.doFilter(request, response);
+        }
+    }
+
+    protected boolean doFilterIdempotent(HttpServletRequest request, HttpServletResponse response, FilterChain chain)
+            throws IOException, ServletException {
+        String method = request.getMethod();
+        if (IDEMPOTENT_METHODS.contains(method)) {
+            log.debug("No idempotent processing done: method is already idempotent: {}", method);
+            return false;
+        }
+        String key = request.getHeader(HEADER_KEY);
+        if (key == null) {
+            log.debug("No idempotent processing done: no {} header present", HEADER_KEY);
+            return false;
+        }
+        log.debug("Idempotent request key: {}", key);
+        KeyValueService kvs = Framework.getService(KeyValueService.class);
+        if (kvs == null) {
+            log.debug("KeyValueService not present");
+            return false;
+        }
+        KeyValueStore store = kvs.getKeyValueStore(getStoreName());
+        String storeStatus = store.getString(key + INFO_SUFFIX);
+        byte[] storeContent = store.get(key);
+        if (storeStatus == null) {
+            log.debug("Handle new request for key: {}", key);
+            long ttl = getTTL().toSeconds();
+            store.put(key + INFO_SUFFIX, INPROGRESS_MARKER, ttl);
+            try {
+                try (CopyingResponseWrapper wrapper = new CopyingResponseWrapper(DEFERRED_OUTPUT_STREAM_THRESHOLD,
+                        response)) {
+                    chain.doFilter(request, wrapper);
+                    wrapper.flushBuffer();
+                    byte[] content = wrapper.getCaptureAsBytes();
+                    wrapper.setHeader(HEADER_KEY, key);
```

Comment:
You won't be able to set the header if the response is committed, which will usually be the case (unless buffering) after the flush. So maybe set before calling `doFilter`?